### PR TITLE
Add spot prices

### PR DESCRIPTION
**Pull request by hidalgopl**
So we need to run https://github.com/banzaicloud/cloudinfo (it scrapes pricing) as a sidecar container and redis to persist it.
By default, scrape is done once per 24h. I've added spot price column to nodeless forecast tab, as well as total price of running only on spot instances to summary.

### CODE REVIEW

**Comment by hidalgopl** [comment]
tagging you guys @jrroman @ldx more as fyi; as it is not urgent, feel free to take a look whenever you have a few minutes.

File: kustomize/overlays/azure/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      - name: nodeless-cost-calculator
+        env:
+        - name: CLOUD_PROVIDER
+          value: azure
+        - name: REGION
+          value: "East US"
+      - name: cloudinfo
+        readinessProbe:
+          httpGet:
+            path: /api/v1/providers/azure/services/compute/regions/eastus/products
```
Comment (hidalgopl) [inline kustomize/overlays/azure/deployment.yaml:21]: this is crucial line. If scrape is now finished, cost-calculator will crash, so we mark cloudinfo container as ready once scraping is finished

File: cost_calculator/instance_selector.py [hunk 2]
```diff
@@ -176,12 +194,69 @@ def get_cheapest_instance(self, cpu_request, memory_request, gpu_spec):
...
+        "gce": "google"
+    }
+
+    def __init__(self, provider, redis_client):
+        self.provider = provider
+        self.redis_client = redis_client
+        self.key_pattern = "/banzaicloud.com/cloudinfo/providers/{provider}/regions/{region}/prices/{instance_type}"
+
+    def _get_azure_region_key(self, region):
+        return region.replace(" ", "").lower()
```
Comment (ldx) [inline cost_calculator/instance_selector.py:220]: Not sure if this is applicable here, but I remember in the Azure instance price scraper we do some cleaning when it comes: to region names: https://github.com/elotl/kip/blob/master/scripts/create_instance_data/create_azure_instance_data.py#L22-L84

File: cost_calculator/instance_selector.py [hunk 3]
```diff
@@ -176,12 +194,69 @@ def get_cheapest_instance(self, cpu_request, memory_request, gpu_spec):
...
+    def _get_data_for_instance(self, instance_type, region):
+        key = self._get_key(region=region, instance_type=instance_type)
+        data = self.redis_client.get(key)
+        prices = self._convert_entry_to_dict(data)
+        return prices
+
+    def get_spot_price(self, instance_type, region):
+        prices = self._get_data_for_instance(instance_type, region)
+        spot_price = self._get_lowest_spot_price(prices)
+        return spot_price
```
Comment (ldx) [inline cost_calculator/instance_selector.py:255]: Does this return a time series of spot prices over time?

File: kustomize/base/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+              mountPath: /etc/cloudinfo/config.toml
+              subPath: config.toml
+              readOnly: true
+          readinessProbe:
+            httpGet:
+              port: 8000
+              path: /api/v1/providers/azure/services/compute/regions/eastus/products
+            initialDelaySeconds: 120
+            periodSeconds: 60
+            timeoutSeconds: 1200
```
Comment (ldx) [inline kustomize/base/deployment.yaml:58]: Some food for thought: here the three services are coupled together; if the deployment is changed or reconfigured, then the whole pod will restart which means all the data in redis will go away, the user will need to wait for cloudinfo to re-populate it, etc. If they are deployed as independent pods instead, then e.g. updating the cost calculator will not affect data in Redis, or trigger a re-scraping of all pricing info (which is time consuming). Probably a better UX if a user updates the app, or changes its configuration.

Comment (hidalgopl) [reply kustomize/base/deployment.yaml:58]: That's a good point. My first take was to have it all as one pod, as initial idea was to share instance data as a file mounted in both containers, but this turn out to be not supported out of the box as data storage. With redis indeed it makes much more sense to have it separate.

File: cost_calculator/instance_selector.py (hunk 3 above)
Comment (hidalgopl) [reply cost_calculator/instance_selector.py:255]: No, it returns current spot price for given instance. Probably in kip we would like to tweak scraping behavior to have more up to date prices.

File: kustomize/base/deployment.yaml (hunk 4 above)
Comment (hidalgopl) [reply kustomize/base/deployment.yaml:58]: addressed it in separate PR: https://github.com/elotl/nodeless-cost-calculator/pull/8 which is based on this branch

**Comment by hidalgopl** [comment]
@ldx @jrroman would like to merge this as well as https://github.com/elotl/nodeless-cost-calculator/pull/8/files . Let me know if you see any outstading issues it those two :)

**Review by ldx — APPROVED** [review]

**Pull request merged**